homeassistant/components/air_pollutants/nilu.py

```diff
@@ -0,0 +1,311 @@
+    import niluclient as nilu
+    name = config.get(CONF_NAME)
+    area = config.get(CONF_AREA)
+    stations = config.get(CONF_STATION)
+    show_on_map = config.get(CONF_SHOW_ON_MAP)
+
+    sensors = []
+
+    if area:
+        stations = nilu.lookup_stations_in_area(area)
+    elif not area and not stations:
+        latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+        longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+        location_client = nilu.create_location_client(latitude, longitude)
+        stations = location_client.station_names
+
+    for station in stations:
+        client = NiluData(nilu.create_station_client(station))
+        client.update()
+        sensors.append(NiluSensor(client, name, show_on_map))
+
+    add_entities(sensors, True)
+
+
+class NiluData:
+    """Class for handling the data retrieval."""
+
+    def __init__(self, api):
+        """Initialize the data object."""
+        self.api = api
+
+    @property
+    def data(self):
+        """Get data cached in client."""
+        return self.api.data
+
+    @Throttle(SCAN_INTERVAL)
+    def update(self):
+        """Get the latest data from nilu apis."""
+        self.api.update()
+
+
+class NiluSensor(AirPollutantsEntity):
+    """Single nilu station air sensor."""
+
+    def __init__(self,
+                 api_data: NiluData,
+                 name: str,
+                 show_on_map: bool):
+        """Initialize the sensor."""
+        self._api = api_data
+        self._name = "{} {}".format(name, api_data.data.name)
+        self._max_aqi = None
+        self._state = None
+        self._attrs = {}
+
+        if show_on_map:
+            self._attrs[CONF_LATITUDE] = api_data.data.latitude
+            self._attrs[CONF_LONGITUDE] = api_data.data.longitude
+
+    @property
+    def attribution(self) -> str:
+        """Return the attribution."""
+        return CONF_ATTRIBUTION
+
+    @property
+    def device_state_attributes(self) -> dict:
+        """Return other details about the sensor state."""
+        return self._attrs
+
+    @property
+    def icon(self) -> str:
+        """Icon to use in the frontend."""
+        return ICON
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self) -> str:
+        """Return the state of the sensor."""
+        import niluclient as nilu
+        if self.air_quality_index \
+                and self.air_quality_index < len(nilu.POLLUTION_INDEX):
+            return nilu.POLLUTION_INDEX[self.air_quality_index]
+
+        return None
+
+    @property
+    def air_quality_index(self) -> str:
+        """Return the Air Quality Index (AQI)."""
+        return self._max_aqi
+
+    @property
+    def carbon_monoxide(self) -> str:
+        """Return the CO (carbon monoxide) level."""
+        from niluclient import CO
+        if CO in self._api.data.sensors:
```

Comment:
Extract the rest of the property code into a helper method that accepts the measurement constant and returns the value or `None`. Almost all the properties use the same pattern.